modules/net/quic/socket.c

```diff
@@ -2385,6 +2416,68 @@ static int quic_sock_get_crypto_secret(struct sock *sk, u32 len,
 	return 0;
 }
 
+static int quic_sock_get_stream_peeloff(struct sock *sk, u32 len,
+					sockptr_t optval, sockptr_t optlen)
+{
+	struct quic_stream_table *streams = quic_streams(sk);
+	struct quic_stream_peeloff peeloff = {};
+	struct quic_frame *frame, *next;
+	struct quic_stream *stream;
+	struct socket *sock;
+	struct file *file;
+	int retval;
+
+	if (len < sizeof(peeloff) || !quic_is_established(sk))
+		return -EINVAL;
+	len = sizeof(peeloff);
+	if (copy_from_sockptr(&peeloff, optval, len))
+		return -EFAULT;
+
+	/* Do not peel off from one netns to another one. */
+	if (!net_eq(current->nsproxy->net_ns, sock_net(sk)))
+		return -EINVAL;
+
+	stream = quic_stream_find(streams, peeloff.stream_id);
+	if (!stream || stream->peeled)
+		return -EINVAL;
+
+	sock = quic_inet_stream_sock_alloc(sk, stream);
+	if (!sock)
+		return -ENOMEM;
+
+	retval = get_unused_fd_flags(peeloff.flags & SOCK_CLOEXEC);
+	if (retval < 0) {
+		sock_release(sock);
+		return retval;
+	}
+
+	file = sock_alloc_file(sock, 0, NULL);
+	if (IS_ERR(file)) {
+		put_unused_fd(retval);
+		return PTR_ERR(file);
+	}
+
+	peeloff.sd = retval;
+	if (peeloff.flags & SOCK_NONBLOCK)
+		file->f_flags |= O_NONBLOCK;
+
+	if (copy_to_sockptr(optlen, &len, sizeof(len)) || copy_to_sockptr(optval, &peeloff, len)) {
+		fput(file);
+		put_unused_fd(retval);
+		return -EFAULT;
+	}
+	/* Move the frames of this stream from inq recv list to its own list. */
+	list_for_each_entry_safe(frame, next, &quic_inq(sk)->recv_list, list) {
+		if (frame->stream != stream)
+			continue;
+		list_move_tail(&frame->list, &stream->recv.frame_list);
+	}
+	sock_hold(sk); /* Keep the parent sk alive while the peeled stream is in use. */
```

Comment:
ah very nice that is already exists. Are there other "parent-child socket" relationships like this already, or does it just exist for other purposes?